specs/stein/approved/amd-sev-libvirt-support.rst

```diff
@@ -0,0 +1,367 @@
+``nova.virt.libvirt.LibvirtConfigGuestFeatureSEV`` class.
+
+This work could also mean that images with the property
+``trait:HW_CPU_x86_SEV=required`` would similarly affect the process
+of launching instances.
+
+Immediate follow-up work
+------------------------
+
+It is suggested that follow-up specs be submitted detailing
+the following subsequent work:
+
+- Support a built-in trait which would be automatically detected per
+  compute host by nova's libvirt driver.  This would most likely be
+  called ``HW_CPU_x86_SEV`` or similar, as an extension of `the
+  existing CPU traits mapping
+  <https://github.com/openstack/nova/blob/c5a7002bd571379818c0108296041d12bc171728/nova/virt/libvirt/utils.py#L47>`_.
+
+  Additional logic would be required to check that the various layers
+  of the hardware and software hypervisor stack are *all* SEV-capable:
+
+  - The presence of the following XML in the response from a libvirt
+    `virConnectGetDomainCapabilities()
+    <https://libvirt.org/html/libvirt-libvirt-domain.html#virConnectGetDomainCapabilities>`_
+    API call `indicates that both QEMU and the AMD Secure Processor
+    (AMD-SP) support SEV functionality
+    <https://libvirt.org/git/?p=libvirt.git;a=commit;h=6688393c6b222b5d7cba238f21d55134611ede9c>`_::
+
+        <domainCapabilities>
+          ...
+          <features>
+            ...
+            <sev supported='yes'/>
+              ...
+            </sev>
+          </features>
+        </domainCapabilities>
+
+  - ``/sys/module/kvm_amd/parameters/sev`` should have the value one
+    to indicate that the kernel has SEV capabilities enabled.
+
+- Since `migration of instances between unencrypted and SEV-encrypted
+  states is not supported in either direction
+  <https://github.com/qemu/qemu/commit/8fa4466d77b44f4f58f3836601f31ca5e401485d>`_,
+  prevent nova from resizing an instance from a non-SEV-enabled flavor
+  to an SEV-enabled one, or vice-versa.  Alternatively, nova could
+  catch the error raised by QEMU, which would be propagated via
+  libvirt, and handle it appropriately.
+
+  Similarly, attempts to suspend / resume a domain should either be
+  prevented, or the error caught and handled.
+
+Alternatives
+------------
+
+#. The idea of initially using a custom trait could be skipped in
+   favour of immediately implementing the built-in trait
+   (``HW_CPU_x86_SEV`` or similar).  This would increase the amount of
+   effort required in order to reach a working prototype, but
+   conversely has the advantage of not requiring the merging of
+   unfinished proof-of-concept code into nova's ``master`` branch.
+
+#. Rather than using a single trait to both facilitate the matching
+   of instances requiring SEV with SEV-capable compute hosts *and*
+   indicate to nova's libvirt driver that SEV should be used when
+   booting ...
+
+   However this would create extra friction for the operators defining SEV-enabled
+   flavors,  it is hard to imagine
```

Comment:
unfinished sentence? leave it out?

Reply:
No, there's an important point here so I'll finish it.